### PR TITLE
Cap stack/service deploy collections

### LINKED ISSUE #2563: server: cap grid_service_deploys collection
Looks like grid_service_deploys collection just keeps growing, currently 280mb (noticed this with https://github.com/matti/kontena-heroku that runs a free mlab instance that has 0.5gb limit)

Shouldn't this collection be capped along with CONTAINER_STATS_CAPPED_SIZE, EVENT_LOGS_CAPPED_SIZE and CONTAINER_LOGS_CAPPED_SIZE

?

### PR DESCRIPTION
**Pull request by jakolehm**
Fixes #2563 . Also adds missing index to the `finished_at` field.

### CODE REVIEW

**Comment by SpComb** [comment]
Hmm, I think I'd prefer to avoid schema changes/migrations like this for the 1.4.x release unless it's absolutely necessary.... capping the stack/service deploy collection does makes plenty of sense, can it wait for 1.5?

OTOH adding a `finished_at` index does sound appropriate for a 1.4.x release.

**Comment by jakolehm** [comment]
Moved index part to #3046

File: server/app/models/grid_service_deploy.rb [hunk 1]
```diff
@@ -36,6 +36,7 @@ class GridServiceDeploy
   index({ created_at: 1 }, { background: true })
   index({ queued_at: 1 }, { background: true })
   index({ started_at: 1 }, { background: true })
+  index({ finished_at: 1 }, { background: true })
```
Comment (SpComb) [inline server/app/models/grid_service_deploy.rb]: The `finished_at` index change probably doesn't belong in this PR... it seems to need some more complex compound index per #3046.

File: server/db/migrations/31_cap_deploys.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
+class CapDeployCollections < Mongodb::Migration
```
Comment (SpComb) [inline server/db/migrations/31_cap_deploys.rb]:
```
api_1      | I, [2017-12-11T10:04:01.278963 #12]  INFO -- Mongodb::Migrator: migrating CapDeploys
api_1      | [44] + Gemfile in context: /app/Gemfile
api_1      | /usr/lib/ruby/gems/2.4.0/gems/activesupport-4.2.8/lib/active_support/inflector/methods.rb:261:in `block in constantize': uninitialized constant CapDeploys (NameError)
api_1      | 	from /usr/lib/ruby/gems/2.4.0/gems/activesupport-4.2.8/lib/active_support/inflector/methods.rb:259:in `each'
api_1      | 	from /usr/lib/ruby/gems/2.4.0/gems/activesupport-4.2.8/lib/active_support/inflector/methods.rb:259:in `inject'
api_1      | 	from /usr/lib/ruby/gems/2.4.0/gems/activesupport-4.2.8/lib/active_support/inflector/methods.rb:259:in `constantize'
api_1      | 	from /usr/lib/ruby/gems/2.4.0/gems/activesupport-4.2.8/lib/active_support/core_ext/string/inflections.rb:66:in `constantize'
api_1      | 	from /app/app/services/mongodb/migration_proxy.rb:23:in `load_migration'
api_1      | 	from /app/app/services/mongodb/migration_proxy.rb:18:in `migration'
api_1      | 	from /app/app/services/mongodb/migration_proxy.rb:4:in `migrate'
api_1      | 	from /app/app/services/mongodb/migrator.rb:90:in `block in migrate_without_lock'
api_1      | 	from /app/app/services/mongodb/migrator.rb:87:in `each'
api_1      | 	from /app/app/services/mongodb/migrator.rb:87:in `migrate_without_lock'
api_1      | 	from /app/app/services/mongodb/migrator.rb:75:in `migrate'
api_1      | 	from config/puma.rb:11:in `block in _load_from'
api_1      | 	from /usr/lib/ruby/gems/2.4.0/gems/puma-3.9.1/lib/puma/configuration.rb:271:in `block in run_hooks'
api_1      | 	from /usr/lib/ruby/gems/2.4.0/gems/puma-3.9.1/lib/puma/configuration.rb:271:in `each'
api_1      | 	from /usr/lib/ruby/gems/2.4.0/gems/puma-3.9.1/lib/puma/configuration.rb:271:in `run_hooks'
api_1      | 	from /usr/lib/ruby/gems/2.4.0/gems/puma-3.9.1/lib/puma/cluster.rb:259:in `worker'
api_1      | 	from /usr/lib/ruby/gems/2.4.0/gems/puma-3.9.1/lib/puma/cluster.rb:126:in `block (2 levels) in spawn_workers'
api_1      | 	from /usr/lib/ruby/gems/2.4.0/gems/puma-3.9.1/lib/puma/cluster.rb:126:in `fork'
api_1      | 	from /usr/lib/ruby/gems/2.4.0/gems/puma-3.9.1/lib/puma/cluster.rb:126:in `block in spawn_workers'
api_1      | 	from /usr/lib/ruby/gems/2.4.0/gems/puma-3.9.1/lib/puma/cluster.rb:122:in `times'
api_1      | 	from /usr/lib/ruby/gems/2.4.0/gems/puma-3.9.1/lib/puma/cluster.rb:122:in `spawn_workers'
api_1      | 	from /usr/lib/ruby/gems/2.4.0/gems/puma-3.9.1/lib/puma/cluster.rb:447:in `run'
api_1      | 	from /usr/lib/ruby/gems/2.4.0/gems/puma-3.9.1/lib/puma/launcher.rb:174:in `run'
api_1      | 	from /usr/lib/ruby/gems/2.4.0/gems/puma-3.9.1/lib/puma/cli.rb:77:in `run'
api_1      | 	from /usr/lib/ruby/gems/2.4.0/gems/puma-3.9.1/bin/puma:10:in `<top (required)>'
api_1      | 	from /usr/bin/puma:23:in `load'
api_1      | 	from /usr/bin/puma:23:in `<main>'
```

Comment (SpComb) [reply server/db/migrations/31_cap_deploys.rb]: Renaming the migration to `31_cap_deploy_collections.rb' fixes it.

**Comment by SpComb** [comment]
Hmm... I think this isn't going to work?

```
api_1      | E, [2017-12-11T10:09:24.259979 #8] ERROR -- StackDeployWorker: failing update: objects in a capped ns cannot grow (10003)
api_1      | E, [2017-12-11T10:09:24.261334 #8] ERROR -- StackDeployWorker: /usr/lib/ruby/gems/2.4.0/gems/mongo-2.4.3/lib/mongo/operation/result.rb:256:in `validate!'
api_1      | /usr/lib/ruby/gems/2.4.0/gems/mongo-2.4.3/lib/mongo/operation/write/write_command_enabled.rb:57:in `execute_write_command'
api_1      | /usr/lib/ruby/gems/2.4.0/gems/mongo-2.4.3/lib/mongo/operation/write/write_command_enabled.rb:41:in `execute'
api_1      | /usr/lib/ruby/gems/2.4.0/gems/mongo-2.4.3/lib/mongo/collection/view/writable.rb:244:in `block in update'
api_1      | /usr/lib/ruby/gems/2.4.0/gems/mongo-2.4.3/lib/mongo/retryable.rb:104:in `write_with_retry'
api_1      | /usr/lib/ruby/gems/2.4.0/gems/mongo-2.4.3/lib/mongo/collection/view/writable.rb:234:in `update'
api_1      | /usr/lib/ruby/gems/2.4.0/gems/mongo-2.4.3/lib/mongo/collection/view/writable.rb:209:in `update_one'
api_1      | /usr/lib/ruby/gems/2.4.0/gems/mongoid-5.2.1/lib/mongoid/query_cache.rb:182:in `update_one_with_clear_cache'
api_1      | /usr/lib/ruby/gems/2.4.0/gems/mongoid-5.2.1/lib/mongoid/persistable.rb:212:in `persist_atomic_operations'
api_1      | /usr/lib/ruby/gems/2.4.0/gems/mongoid-5.2.1/lib/mongoid/persistable.rb:195:in `persist_or_delay_atomic_operation'
api_1      | /usr/lib/ruby/gems/2.4.0/gems/mongoid-5.2.1/lib/mongoid/persistable.rb:146:in `prepare_atomic_operation'
api_1      | /usr/lib/ruby/gems/2.4.0/gems/mongoid-5.2.1/lib/mongoid/persistable/settable.rb:23:in `set'
api_1      | /app/app/jobs/stack_deploy_worker.rb:53:in `deploy_service'
api_1      | /app/app/jobs/stack_deploy_worker.rb:27:in `block in deploy_stack'
api_1      | /app/app/jobs/stack_deploy_worker.rb:25:in `each'
api_1      | /app/app/jobs/stack_deploy_worker.rb:25:in `deploy_stack'
api_1      | /app/app/jobs/stack_deploy_worker.rb:12:in `perform'
api_1      | /usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/calls.rb:28:in `public_send'
api_1      | /usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/calls.rb:28:in `dispatch'
api_1      | /usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/call/sync.rb:16:in `dispatch'
api_1      | /usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/cell.rb:50:in `block in dispatch'
api_1      | /usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/cell.rb:76:in `block in task'
api_1      | /usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/actor.rb:339:in `block in task'
api_1      | /usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/task.rb:44:in `block in initialize'
api_1      | /usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/task/fibered.rb:14:in `block in create'
```

**Comment by SpComb** [comment]
It looks like grid service deploys are also failing silently... adding some logging, it looks like the [`fetch_deploy_item` => `find_one_and_update`](https://github.com/kontena/kontena/blob/e86182973309ec25c0c1120c3bcbe2a33fc2f7a6/server/app/jobs/grid_service_scheduler_worker.rb#L56-L62) to change the `queued_at` is failing:

```
api_1      | W, [2017-12-11T10:16:27.756392 #9]  WARN -- GridServiceSchedulerWorker: exception: failing update: objects in a capped ns cannot grow (10003) (Mongo::Error::OperationFailure)
api_1      | /usr/lib/ruby/gems/2.4.0/gems/mongo-2.4.3/lib/mongo/operation/result.rb:256:in `validate!'
api_1      | /usr/lib/ruby/gems/2.4.0/gems/mongo-2.4.3/lib/mongo/operation/executable.rb:37:in `block in execute'
api_1      | /usr/lib/ruby/gems/2.4.0/gems/mongo-2.4.3/lib/mongo/server/connection_pool.rb:107:in `with_connection'
api_1      | /usr/lib/ruby/gems/2.4.0/gems/mongo-2.4.3/lib/mongo/server.rb:247:in `with_connection'
api_1      | /usr/lib/ruby/gems/2.4.0/gems/mongo-2.4.3/lib/mongo/operation/executable.rb:35:in `execute'
api_1      | /usr/lib/ruby/gems/2.4.0/gems/mongo-2.4.3/lib/mongo/collection/view/writable.rb:118:in `block in find_one_and_update'
api_1      | /usr/lib/ruby/gems/2.4.0/gems/mongo-2.4.3/lib/mongo/retryable.rb:104:in `write_with_retry'
api_1      | /usr/lib/ruby/gems/2.4.0/gems/mongo-2.4.3/lib/mongo/collection/view/writable.rb:111:in `find_one_and_update'
api_1      | /usr/lib/ruby/gems/2.4.0/gems/mongoid-5.2.1/lib/mongoid/query_cache.rb:182:in `find_one_and_update_with_clear_cache'
api_1      | /usr/lib/ruby/gems/2.4.0/gems/mongoid-5.2.1/lib/mongoid/contextual/mongo.rb:189:in `find_one_and_update'
api_1      | /usr/lib/ruby/gems/2.4.0/gems/mongoid-5.2.1/lib/mongoid/contextual.rb:20:in `find_one_and_update'
api_1      | /app/app/jobs/grid_service_scheduler_worker.rb:59:in `fetch_deploy_item'
api_1      | /app/app/jobs/grid_service_scheduler_worker.rb:23:in `check_deploy_queue'
api_1      | /app/app/jobs/grid_service_scheduler_worker.rb:12:in `block in watch'
api_1      | /app/app/jobs/grid_service_scheduler_worker.rb:11:in `loop'
api_1      | /app/app/jobs/grid_service_scheduler_worker.rb:11:in `watch'
api_1      | /usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/calls.rb:28:in `public_send'
api_1      | /usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/calls.rb:28:in `dispatch'
api_1      | /usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/call/async.rb:7:in `dispatch'
api_1      | /usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/cell.rb:50:in `block in dispatch'
api_1      | /usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/cell.rb:76:in `block in task'
api_1      | /usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/actor.rb:339:in `block in task'
api_1      | /usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/task.rb:44:in `block in initialize'
api_1      | /usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/task/fibered.rb:14:in `block in create'
```

**Comment by jakolehm** [comment]
Ok, this was a bad idea.
